File: sunflower-core/src/main/java/org/tdf/sunflower/state/AbstractStateTrie.java
package org.tdf.sunflower.state;

import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.tdf.common.store.DatabaseStore;
import org.tdf.common.store.Store;
import org.tdf.common.trie.ReadOnlyTrie;
import org.tdf.common.trie.Trie;
import org.tdf.common.util.HexBytes;


@Slf4j(topic = "trie")
public abstract class AbstractStateTrie<ID, S> implements StateTrie<ID, S> {
    abstract Store<byte[], byte[]> getDB();

    public S get(HexBytes rootHash, ID id) {
        return getTrieForReadOnly(rootHash).get(id);
    }


    public Trie<ID, S> getTrie(HexBytes rootHash) {
        return getTrie().revert(rootHash);
    }

    @SneakyThrows
    protected Trie<ID, S> getTrieForReadOnly(HexBytes rootHash) {
        return ReadOnlyTrie.of(getTrie(rootHash));
    }

}

File: sunflower-core/src/main/java/org/tdf/sunflower/state/AbstractStateTrie.kt
package org.tdf.sunflower.state

import lombok.SneakyThrows
import org.slf4j.LoggerFactory
import org.tdf.common.store.Store
import org.tdf.common.trie.ReadOnlyTrie
import org.tdf.common.trie.Trie
import org.tdf.common.util.HexBytes

abstract class AbstractStateTrie<ID, S> : StateTrie<ID, S> {
    abstract val db: Store<ByteArray, ByteArray>

    override fun get(rootHash: HexBytes, id: ID): S? {
        return getTrieForReadOnly(rootHash)[id]
    }

    override fun getTrie(rootHash: HexBytes): Trie<ID, S> {
        return trie.revert(rootHash)
    }

    @SneakyThrows
    protected fun getTrieForReadOnly(rootHash: HexBytes): Trie<ID, S> {
        return ReadOnlyTrie.of(getTrie(rootHash))
    }

    companion object {
        private val log = LoggerFactory.getLogger("trie")
    }
}